### ISSUE TITLE
Low Level Http2 Secured Configuration

### ISSUE THREAD

**Issue by amis0405** [eclipse/jetty.project#6177]
**Jetty version**
11.02
**Java version**
java 11
**Question**
I have Implemented Low level HTTP2 client-server at localhost. I want to make it secured only connection. In High level implementation we can use TLS keystores and certificates to do the same but how to do that in low-level API's ? Is such thing even possible/necessary ?

Secondly is there a way to configure for Asynchronus/Synchronus request and Blocking/Non-Blocking I/O at Low-Level ?

@sbordet

**Comment by joakime** [comment]
Please explain what you mean by "low level" and "high level"?
As your description is confusing.

From Jetty's point of view, the ServerConnector is where HTTP/2 is configured for the server (there is no high level or low level distinction).
The jetty-client in 11.0.2 is also capable of supporting both HTTP/1 and HTTP/2 with the same configuration (there is also no high level or low level distinction in the client as well).

In both cases, server and client, there is a SslContextFactory configuration (either `SslContextFactory.Server` or `SslContextFactory.Client` depending on role), this is where your SSL/TLS level configuration exists.
The `SslContextFactory`, in turn, merely configures the Java JVM's `javax.net.ssl.SSLEngine`, which internally creates and manages the `javax.net.ssl.SSLSession` that is used for the actual connection.  Jetty merely tells Java its desired configuration, Java does all the heavy lifting with SSL/TLS and its encryption.
Whether the connection (client or server) requires a TLS keystore or not is dependent on many things about your specific scenario. (this is far too complex of a topic for a simple github issue)

See documentation for more:
* https://www.eclipse.org/jetty/documentation/jetty-11/operations-guide/index.html#og-protocols-http2
* https://www.eclipse.org/jetty/documentation/jetty-11/operations-guide/index.html#og-keystore

> Secondly is there a way to configure for Asynchronus/Synchronus request and Blocking/Non-Blocking I/O at Low-Level ?

Jetty is 100% async and non-blocking.

However we have to honor the various API contracts that various web applications choose, so we mimic blocking for the API calls that require it (but this mimicing results in a blocked thread, not any kind of alternate I/O technique)
Look at the Jakarta EE 9 Servlet API async features (Tip: async processing and async I/O are different parts of the spec).
Even the Jakarta EE 9 Websocket API has async and blocking API calls.

**Comment by amis0405** [comment]
The low-level HTTP/2 support is provided by org.eclipse.jetty.http2.server.RawHTTP2ServerConnectionFactory and org.eclipse.jetty.http2.api.server.ServerSessionListener for SERVER and org.eclipse.jetty.http2.client.HTTP2Client for CLIENT.That what I meant by Low-Level API's.

So you are saying that the Low-Level API's must be configured the same way as High level ones using 
> ServerConnector secureConnector = new ServerConnector(server, tls, alpn, rawHTTP2ServerConnectionFactory);

on Server and client equivalent on Client side  for Secured HTTP2(H2) ?

**Comment by joakime** [comment]
> The low-level HTTP/2 support is provided by org.eclipse.jetty.http2.server.RawHTTP2ServerConnectionFactory and org.eclipse.jetty.http2.api.server.ServerSessionListener for SERVER

With that you have no traditional Request or Response objects, no Handler support, no Servlet support, no Webapp support, no distinction for Asynchronus/Synchronus or Blocking/Non-Blocking.
That's purely the HTTP/2 protocol and that's it, you only talk to that layer in terms of protocol objects and raw buffers.

We only use the `RawHTTP2ServerConnectionFactory` ourselves in unit test cases, typically to test corner cases of the HTTP/2 spec (errors, bad behaviors, failure scenarios, etc) that are tricky to trigger in a consistent way.

Note that if you want to use Handlers, Servlets, WebSocket, WebApps, etc then you will want to use the traditional `HTTP2ServerConnectionFactory` instead.

> ``` java
> ServerConnector secureConnector = new ServerConnector(server, tls, alpn, rawHTTP2ServerConnectionFactory);
> ```

That `RawHTTP2ServerConnectionFactory` should operate as all of the other `ConnectionFactory` instances on the `ServerConnector`, just setup your ssl/tls and alpn layers before it (like you have) and you should have SSL/TLS enabled for that connector.  (Make sure you configure your `SslConnectionFactory` and `ALPNServerConnectionFactory` properly.)

> on Server and client equivalent on Client side for Secured HTTP2(H2) ?

The client side has it's own `SslContextFactory.Client` for its configuration of SSL/TLS to the JVM, you may have to alter the default configuration (but that depends on what you are doing)

**Comment by sbordet** [comment]
@amis0405 you don't need to use TLS if you don't want to. HTTP/2 does not mandate use of TLS. Only certain client implementations (all browsers) have decided to use HTTP/2 only over TLS; Jetty, both on client and on server, does not require TLS when you use HTTP/2.

All the low-level APIs are non-blocking.

**Comment by amis0405** [comment]
What I want is to have only a secured(https) connection so is TLS implementation necessary to achieve this ? As far as I know connection will be secured only if we are using certificates/keystores...

> All the low-level APIs are non-blocking.

Thats means Asynchronous right ?

**Comment by sbordet** [comment]
If you want TLS you have to use `SslContextFactory.[Client|Server]` like @joakime explained.

>> All the low-level APIs are non-blocking.

> Thats means Asynchronous right ?

No, that means non-blocking. "Asynchronous" is typically related but not the same as "non-blocking".
The fact that you ask means that probably your definition of "asynchronous" is not the same as ours, so you have to clarify what you exactly means and why it is important to you.

**Comment by amis0405** [comment]
Okay let not get into complex discussions of Asynchronous.. What i meant is one should get immediate response... it does sounds like non-blocking

**Comment by amis0405** [comment]
Even after doing the TLS setup(SslContextFactory etc) on both the sides(which works for Higher API's)  the request-response calls are not flowing in LOW LEVEL APIs.
On Server side "onAccept" is called and on Client side "onPreface" is called and then request is not fired, which actually works if I configure it without TLS.What could be the issue here ?

**Comment by sbordet** [comment]
@amis0405 it's not clear what you're doing, what do you expect, and what you actually get.

**Comment by amis0405** [comment]
I have implement a client-server(Low Level APIs) at localhost to send and receive JSON data. The implementation works fines with request-response using Session.Listener exchanging JSON data.But when I try to make this as SECURED ONLY(H2) over TLS using the necessary settings like `SslContextFactory.[Client|Server]` , the request-response methods are not triggered. Though the client-server are started and connected, the only methods that are triggered are "onAccept" on Server side and "onPreface" on Client side and no further processing takes place. It behaves just like when the server(LL) is accessed from any browser. What could be the issue or possible fix?
PS. The SslContextFactory settings I have used is working for my HIGH LEVEL API implementation of client-server so basically no issues with the TLS config there.

**Comment by sbordet** [comment]
@amis0405 we have an array of test cases similar to yours that prove that the HTTP/2 implementation is working fine.

You have yet to provide evidence of your problem in form of code or DEBUG logs, so we cannot help you.
Chances are that you are doing something wrong in your code or setup and it is not a Jetty issue.

**Comment by joakime** [comment]
@amis0405 the information we have requested to troubleshoot this report has not yet been provided.
We cannot help you unless you give us the information we asked for.

**Comment by amis0405** [comment]
[Http2.zip](https://github.com/eclipse/jetty.project/files/6350678/Http2.zip)
Here attached are the client-server implemented example. Kindly let me know if you need any more info on this.
JDK 11

**Comment by sbordet** [comment]
Change this:

```java
CompletableFuture<Session> sessionCF = http2Client.connect(serverAddress, new Session.Listener.Adapter() {
```

to this:

```java
CompletableFuture<Session> sessionCF = http2Client.connect(connector.getSslContextFactory(), serverAddress, new Session.Listener.Adapter() {
```

**Comment by amis0405** [comment]
Thank you both its working now.... My other query was can I not use ALPN over TLS and still make it run ?
I will be only using H2 over TLS (Client/server) so I dont need any negotiations for protocol to be used. So can I  configure server and client to not use any ALPN libraries and still work over TLS(H2) ? 

PS: Probably I have removed all ALPN references from  the above code for SERVER and its running but same cant be said about CLIENT.

**Comment by sbordet** [comment]
@amis0405 currently `HTTP2Client` hard-codes the use of ALPN.

We have made ALPN optional in `HttpClientTransportOverHTTP2` in #441, but not in `HTTP2Client`.

@amis0405 please close this issue as solved, and open a new one for porting #441 to `HTTP2Client`.
Thanks!
